### PR TITLE
Improved version of bitfield generation and reporting

### PR DESCRIPTION
**Pull request by dkfellows**
Tricky point: we were not previously correct about how we were doing the read-back in some cases. That's now all shared code so there's a better chance that we'll do the critical bits (that read structures off the machine) correctly.

### CODE REVIEW

**Review by andrewgait — APPROVED** [review]
Looks fine to me, leaving for @alan-stokes to look at too.

File: spynnaker/pyNN/extra_algorithms/on_chip_bit_field_generator.py [hunk 1]
```diff
@@ -118,288 +136,228 @@ def __call__(
...
-                default_report_folder, self._BIT_FIELD_SUMMARY_REPORT_FILENAME)
-
-    def _read_back_and_summarise_bit_fields(
-            self, app_graph, transceiver, placements,
-            default_report_folder, bit_field_summary_report_name):
+            self._summary_report_bit_fields(app_graph, os.path.join(
+                default_report_folder,
+                self._BIT_FIELD_SUMMARY_REPORT_FILENAME))
+
+    _PER_CORE_SUMMARY = (
```
Comment (alan-stokes) [inline spynnaker/pyNN/extra_algorithms/on_chip_bit_field_generator.py:154]: shouldnt these be at the top of the class?

File: spynnaker/pyNN/extra_algorithms/on_chip_bit_field_generator.py [hunk 2]
```diff
@@ -118,288 +136,228 @@ def __call__(
...
+        for master_pop_key, n_bits, bitfield in self.__bitfields(placement):
+            # put into report
+            for neuron_id in range(n_bits):
+                f.write("for key {}, neuron id {} has bit == {}\n".format(
+                    master_pop_key, neuron_id,
+                    self.__bit_for_neuron_id(bitfield, neuron_id)))
 
-        # read in each bitfield
-        for _bit_field_index in range(0, n_bit_field_entries):
+    def __bit_for_neuron_id(self, bit_field, neuron_id):
```
Comment (alan-stokes) [inline spynnaker/pyNN/extra_algorithms/on_chip_bit_field_generator.py:244]: can be a static method

File: spynnaker/pyNN/extra_algorithms/on_chip_bit_field_generator.py [hunk 3]
```diff
@@ -118,288 +136,228 @@ def __call__(
...
+
+        # read in each bitfield
+        for _bit_field_index in range(total):
+            # master pop key, n words and read pointer; filter_info_t
+            master_pop_key, n_words, read_pointer = _THREE_WORDS.unpack(
+                self.__txrx.read_memory(
+                    placement.x, placement.y, address, _THREE_WORDS.size))
+            address += _THREE_WORDS.size
+            # Mask off merged and all_ones flag bits
+            n_words &= 0x3FFFFFFF
```
Comment (alan-stokes) [inline spynnaker/pyNN/extra_algorithms/on_chip_bit_field_generator.py:293]: magic number

Comment (dkfellows) [reply spynnaker/pyNN/extra_algorithms/on_chip_bit_field_generator.py:293]: Very much so!

File: spynnaker/pyNN/extra_algorithms/on_chip_bit_field_generator.py (hunk 2 above)
Comment (dkfellows) [reply spynnaker/pyNN/extra_algorithms/on_chip_bit_field_generator.py]: Moved to just being a function.

**Comment by dkfellows** [comment]
The changes to the network graphing are just to make things a bit less contorted (and IDs are converted to strings earlier).

The changes to the bitfield compression stuff are mainly to do with finding easier ways to iterate over the work that needs to be done. Iterating over placements is easier than iterating over cores looking for those that have a placement. (The overall amount of work should be similar when a network is mostly filling its assigned machine and most of the cores are neuron cores.)

**Review by alan-stokes — APPROVED** [review]

**Comment by dkfellows** [comment]
[recooking the integration tests…]

**Pull request merged**